### PR TITLE
[SPARK-24098][SQL] ScriptTransformationExec should wait process exiting before output iterator finish

### PR DESCRIPTION
**Pull request by liutang123**
When feed thread doesn't set its _exception variable and the progress doesn't exit completely, output Iterator will return false in hasNext function.

## What changes were proposed in this pull request?
wait script process exiting before output iterator finish.

## How was this patch tested?
manual test:

1. Add Thread.sleep(1000 * 600) before assign for _exception.
2. structure a python script witch will throw exception like follow:
test.py
```import sys 
for line in sys.stdin:   
  raise Exception('error') 
  print line
```
3. use script created in step 2 in transform.
```spark-sql>add files /path_to/test.py;```
```spark-sql>select transform(id) using 'python test.py' as id from city;```
The result is that spark will end successfully.

### CODE REVIEW

**Comment by liutang123** [comment]
Bug Reappearance:
1. Add Thread.sleep(1000 * 600) before assign for _exception.
2. structure a python script witch will throw exception like follow:
test.py
```import sys 
for line in sys.stdin:   
  raise Exception('error') 
  print line
```
3. use script created in step 2 in transform.
```spark-sql>add files /path_to/test.py;```
```spark-sql>select transform(id) using 'python test.py' as id from city;```
The result is that spark will end successfully.

**Comment by liutang123** [comment]
@rxin hi, Do you have time to look at this PR?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala [hunk 1]
```diff
@@ -137,13 +137,12 @@ case class ScriptTransformationExec(
...
           }
 
-          if (!proc.isAlive) {
-            val exitCode = proc.exitValue()
-            if (exitCode != 0) {
-              logError(stderrBuffer.toString) // log the stderr circular buffer
-              throw new SparkException(s"Subprocess exited with status $exitCode. " +
-                s"Error: ${stderrBuffer.toString}", cause)
-            }
+          proc.waitFor()
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala]: basically you are moving the `proc.waitFor()` to after `if (writerThread.exception.isDefined) ...`, why is it?

Comment (liutang123) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala]: Although writerThread._exception is a volatile variable, some times writerThread.exception function may be called before writerThread._exception's assignment due to readThread and writerThread are working parallel.

**Comment by cloud-fan** [comment]
cc @dongjoon-hyun @gatorsmile

**Comment by liutang123** [comment]
@gatorsmile Could you please give some comments when you have time? Thanks so much.
In addition, I think this is a critical bug!!!

**Comment by cxzl25** [comment]
@liutang123  cc @cloud-fan @gatorsmile 

I also encountered this problem.

image
python:
```python
import sys
for line in sys.stdin:
    print 1/0
```
sql:
```sql
ADD FILE test.py;
SELECT TRANSFORM(1) USING 'python test.py'
AS (c1)
```

I solved it this way:  *writerThread.join()*
```java
if (scriptOutputReader.next(scriptOutputWritable) <= 0) {
    writerThread.join()
    checkFailureAndPropagate()
    return false
}
```

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala [hunk 2]
```diff
@@ -174,7 +173,6 @@ case class ScriptTransformationExec(
                     // Ideally the proc should *not* be alive at this point but
                     // there can be a lag between EOF being written out and the process
                     // being terminated. So explicitly waiting for the process to be done.
-                    proc.waitFor()
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala]: We need to update the above comment as well

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala]: You didn't answer my question. It looks more natural to wait for the process then call `writerThread.exception` according to what you said.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala (hunk 2 above)
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala]: I do not think we should remove proc.waitFor() here.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala (hunk 1 above)
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala]: I tested with provided python script. Actually at ScriptTransformationWriterThread, no exception is thrown. 

And, checkFailureAndPropagate is called not due to EOFException, it is called at:

https://github.com/apache/spark/blob/8a17d26784c53fb50b6373b566aab71135c8956f/sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala#L163-L167

In this case, checkFailureAndPropagate can be called before proc is being terminated due to possible lag, so the !proc.isAlive check fails.

I think the fix is to add proc.waitFor() before other checkFailureAndPropagate calls.

Comment (liaoaoyuan97) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala:140]: @viirya In the proposed PR, proc.waitFor() is called within function checkFailureAndPropagate. Basically, proc is terminated before any checkFailureAndPropagate call returns. Do you mean that we should not waitfor proc before other CheckFailureAndPropagate calls but the case you mentioned?

**Pull request closed without merging**